### PR TITLE
Ricl rpc

### PR DESCRIPTION
**Pull request by lzl124631x**
# New APIs

- [x] hmy_newFilter
- [x] hmy_newBlockFilter
- [x] hmy_newPendingTransactionFilter
- [x] hmy_uninstallFilter

Testing:

- [ ] hmy_getLogs
- [ ] hmy_getFilterChanges
  - [x] hmy_newBlockFilter
  - [ ] hmy_newPendingTransactionFilter
- [ ] hmy_getFilterLogs
- [ ] hmy_pendingTransactions (NOT subscription related API)

# Changes
* Code reorg for better mapping from ours to eth. Easier for taking reference.
  * Split APIBackend and Backend.
  * Harmony object (similar to Ethereum Object)

# TODO
- [ ] Add go test to test without web3.

### CODE REVIEW

File: core/resharding.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ const (
 	// GenesisShardNum is the number of shard at genesis
 	GenesisShardNum = 4
 	// GenesisShardSize is the size of each shard at genesis
-	GenesisShardSize = 50
```
Comment (lzl124631x) [inline core/resharding.go]: undo

File: hmy/api_backend.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/harmony-one/harmony/core/state"
+	"github.com/harmony-one/harmony/core/types"
+)
+
+// APIBackend An implementation of Backend. Full client.
+type APIBackend struct {
+	hmy *Harmony
+}
+
+// ChainDb ...
```
Comment (mikedoan) [inline hmy/api_backend.go]: ek in another PR complains about this type of comments. can you write all full comments? i totally agree in our code base there are other similar comments.

File: node/service.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/harmony-one/harmony/accounts"
+)
+
+// ServiceContext is a collection of service independent options inherited from
+// the protocol stack, that is passed to all constructors to be optionally used;
+// as well as utility methods to operate on the service environment.
+type ServiceContext struct {
```
Comment (mikedoan) [inline node/service.go]: where are we using this?

Comment (lzl124631x) [reply node/service.go]: seems like remnant. deleted.

File: hmy/bloombits.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	bloomRetrievalBatch = 16
+
+	// bloomRetrievalWait is the maximum time to wait for enough bloom bit requests
+	// to accumulate request an entire batch (avoiding hysteresis).
+	bloomRetrievalWait = time.Duration(0)
+)
+
+// startBloomHandlers starts a batch of goroutines to accept bloom bit database
+// retrievals from possibly a range of filters and serving the data to satisfy.
+func (hmy *Harmony) startBloomHandlers(sectionSize uint64) {
```
[the remark this answers is not in the document]
Comment (lzl124631x) [reply hmy/bloombits.go]: bloomIndexer related things, part of Filter work.

File: hmy/api_backend.go (hunk 2 above)
Comment (lzl124631x) [reply hmy/api_backend.go]: Will fix later

**Pull request merged**